pymeasure/instruments/agilent/agilentE5062A.py

```diff
@@ -0,0 +1,540 @@
+#
+# This file is part of the PyMeasure package.
+#
+# Copyright (c) 2013-2024 PyMeasure Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+from pymeasure.instruments import Instrument, SCPIMixin, Channel
+from pymeasure.instruments.validators import strict_range, strict_discrete_set
+
+import numpy as np
+
+DISPLAY_LAYOUT_OPTIONS = [
+        "D1",
+        "D12",
+        "D1_2",
+        "D112",
+        "D1_1_2",
+        "D123",
+        "D1_2_3",
+        "D12_33",
+        "D11_23",
+        "D13_23",
+        "D12_13",
+        "D1234",
+        "D1_2_3_4",
+        "D12_34"
+]
+
+
+class AgilentE5062A(SCPIMixin, Instrument):
+    """Represents the Agilent E5062A Vector Network Analyzer
+
+    This VNA has 4 separate channels, each of which has its own sweep. The
+    channels are stored in the ``channels`` dictionary. Channels can be enabled
+    even if not displayed. Reading out data happens at the channel level.
+
+    Each channel can display up to 4 traces (controlled via the ``active_traces``
+    parameter). Note that each channel also has a ``display_layout`` property
+    that controls the layout of the traces in the channel. The traces are
+    accessed via the ``traces`` dictionary (i.e. ``vna.channels[1].traces[1]``).
+
+    The VNA supports multiple transfer formats. This API only supports the IEEE
+    64-bit floating point format. The VNA is configured for this format during
+    initialization, and the data transfer format is not a publicly accessible
+    property.
+
+    This class implements only a subset of the total E5062A functionality. The
+    more significant missing functionality includes (TODO):
+
+    - editing the scale of the graphs
+    - markers
+    - calibration
+    - power sweeps
+
+    """
+
+    class VNAChannel(Channel):
```

Comment:
Interesting solution to make the Channel an attribute of the instrument class.
Does it show in the api documentation? (you can in the docs folder and execute `make html` to create the documentation locally, which will be in the folder `builds`)

Reply:
The classes were showing up in the documentation, but they were missing the nice blue header, so I moved them outside. It seems that the docstring tests also did not check the methods of those classes while they were defined as attributes

Reply:
Ah yes, the tests only find top level classes.